nomenclature/processor/generic.py

```diff
@@ -0,0 +1,153 @@
+import logging
+from collections import Counter
+from pathlib import Path
+
+import yaml
+from pyam import IamDataFrame
+from pydantic import BaseModel, field_validator, ValidationInfo
+from pydantic.types import FilePath
+from pydantic_core import PydanticCustomError
+
+from nomenclature.definition import DataStructureDefinition
+from nomenclature.error import custom_pydantic_errors
+from nomenclature.processor import Processor
+from nomenclature.processor.utils import get_relative_path
+
+logger = logging.getLogger(__name__)
+
+here = Path(__file__).parent.absolute()
+
+
+class AggregationItem(BaseModel):
+    """Item used for aggregation-mapping"""
+
+    name: str
+    components: list[str]
+
+
+class Aggregator(Processor):
+    """Aggregation or renaming of an IamDataFrame on a `dimension`"""
+    file: FilePath
+    dimension: str
+    mapping: list[AggregationItem]
+
+    def apply(self, df: IamDataFrame) -> IamDataFrame:
+        """Apply region processing
+
+        Parameters
+        ----------
+        df : IamDataFrame
+            Input data that the region processing is applied to
+
+        Returns
+        -------
+        IamDataFrame:
+            Processed data
+
+        """
+        return df.rename(
+            mapping={self.dimension: self.rename_mapping},
+            check_duplicates=False,
+        )
+
+    @property
+    def rename_mapping(self):
+        rename_dict = {}
+
+        for item in self.mapping:
+            for c in item.components:
+                rename_dict[c] = item.name
+
+        return rename_dict
+
+    @field_validator("mapping")
+    def validate_target_names(cls, v, info: ValidationInfo):
+        _validate_items([item.name for item in v], info, "Duplicate target")
+        return v
+
+    @field_validator("mapping")
+    def validate_components(cls, v, info: ValidationInfo):
+        # components have to be unique for creating rename-mapping (component -> target)
+        all_components = list()
+        for item in v:
+            all_components.extend(item.components)
+        _validate_items(all_components, info, "Duplicate component")
+        return v
+
+    @field_validator("mapping")
+    def validate_target_vs_components(cls, v, info: ValidationInfo):
+        # guard against having identical target and component
+        _codes = list()
+        for item in v:
+            _codes.append(item.name)
+            _codes.extend(item.components)
+        _validate_items(_codes, info, "Non-unique target and component")
+        return v
+
+    @property
+    def codes(self):
+        _codes = list()
+        for item in self.mapping:
+            _codes.append(item.name)
+            _codes.extend(item.components)
+        return _codes
+
+    def validate_with_definition(self, dsd: DataStructureDefinition) -> None:
+        error = None
+        # check for codes that are not defined in the codelists
+        codelist = getattr(dsd, self.dimension, None)
```

Comment:
I'd let the error occur right here if the dimension is not found:
```suggestion
        codelist = getattr(dsd, self.dimension)
```

Reply:
I like the current implementation better because this way, I can show the name of the offending file directly in the error message, making it easier to debug in repositories that have many files.

Reply:
Fair point, but in this case the error does not really have anything to do with the aggregator processor but rather with the missing dimension in the DataStructueDefinition.